module/zfs/dsl_scan.c

```diff
@@ -403,6 +423,56 @@ dsl_scan_cancel(dsl_pool_t *dp)
 	    dsl_scan_cancel_sync, NULL, 3, ZFS_SPACE_CHECK_RESERVED));
 }
 
+/*
+ * Set scrub pause/resume state if it makes sense to do so
+ */
+int
+dsl_scrub_set_pause_resume(const dsl_pool_t *dp, pool_scrub_cmd_t cmd)
+{
+	spa_t *spa = dp->dp_spa;
+	dsl_scan_t *scn = dp->dp_scan;
+
+	ASSERT(cmd == POOL_SCRUB_NORMAL || cmd == POOL_SCRUB_PAUSE);
+
+	/* can't pause a scrub when there is no in-progress scrub */
+	if (cmd == POOL_SCRUB_PAUSE && !dsl_scan_scrubbing(dp))
+		return (SET_ERROR(ENOENT));
```

Comment:
I think this function would be easier to follow if it was overall structured like:

```
if (cmd == PAUSE) {
  // do checks
  // set fields
} else {
  // ASSERT(cmd == NORMAL)
  if (paused_scrub()) {
    // set fields
  }
}
```